staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go

```diff
@@ -34,41 +34,111 @@ func MaxLength(_ operation.Context, fldPath *field.Path, value, _ *string, max i
 	return nil
 }
 
-// Required verifies that the specified value is not the zero-value for its
-// type.
-func Required[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
+// RequiredValue verifies that the specified value is not the zero-value for
+// its type.
+func RequiredValue[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
+	var zero T
+	if *value != zero {
+		return nil
+	}
+	return field.ErrorList{field.Required(fldPath, "")}
+}
+
+// RequiredPointer verifies that the specified pointer is not nil.
+func RequiredPointer[T any](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
 	if value != nil {
-		var zero T
-		if *value != zero {
-			return nil
-		}
+		return nil
+	}
+	return field.ErrorList{field.Required(fldPath, "")}
+}
+
+// RequiredSlice verifies that the specified slice is not empty.
+func RequiredSlice[T any](_ operation.Context, fldPath *field.Path, value, _ []T) field.ErrorList {
+	if len(value) > 0 {
+		return nil
 	}
 	return field.ErrorList{field.Required(fldPath, "")}
 }
 
-// Forbidden verifies that the specified value is the zero-value for its
+// RequiredMap verifies that the specified map is not empty.
+func RequiredMap[K comparable, T any](_ operation.Context, fldPath *field.Path, value, _ map[K]T) field.ErrorList {
+	if len(value) > 0 {
+		return nil
+	}
+	return field.ErrorList{field.Required(fldPath, "")}
+}
+
+// RequiredValue verifies that the specified value is the zero-value for its
 // type.
-func Forbidden[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
+func ForbiddenValue[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
+	var zero T
+	if value == nil || *value == zero {
```

Comment:
Why the nil check here?  This should never be called with a nil ptr, same as required and optional

Reply:
Actually it could be, in forbidden's unittest
```golang
errs = append(errs,
		func(obj, oldObj *T2, fldPath *field.Path) (errs field.ErrorList) {
			if e := validate.ForbiddenPointer(opCtx, fldPath, obj, oldObj); len(e) != 0 {
				errs = append(errs, e...)
				return // do not proceed
			}
			errs = append(errs, Validate_T2(opCtx, obj, oldObj, fldPath)...)
			return
		}(obj.PT2, safe.Field(oldObj, func(oldObj *T1) *T2 { return oldObj.PT2 }), fldPath.Child("pt2"))...)
```
When `Validate_T2`, it maybe handle a nil obj.

I think we need another way of shortcircuit to stop validating when this current validation pass. e.g. if `ForbiddenPointer` passes then there is no point to validate its typedef (Validate_T2)

Reply:
For required and optional, we would only continue when the Pointer is non-nil

Reply:
When would it ever be nil?  If the input was a pointer, we call the SomethingPointer() function.  If  we call this, it was because the original field was NOT a pointer/slice/map, so we took the address of it.  The address of a field can never be nil.

Reply:
The original field is a pointer, and after ForbiddenPointer() function, it continues to validate a typedef (`Validate_T2` in this case).

Reply:
I see.  So, IIUC, we don't need a nil-check in `validate.ForbiddenValue()` (which is only called on value fields) and `validate.ForbiddenPointer()` is basically JUST a nil check.

But what's weird is that we then proceed to do more validation after proving that the pointer is nil.

What if `forbidden` always emits both forbidden AND optional.  That way, if the pointer is nil, forbidden will pass (no error) and optional will short-circuit.  If the pointer is not nil, forbidden will fail and short-circuit.  Again, I am not REALLY sure what forbidden is for yet.

Reply:
```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go b/staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go
index b94fedc7157..3d81eb3bd44 100644
--- a/staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go
+++ b/staging/src/k8s.io/apimachinery/pkg/api/validate/schema.go
@@ -84,7 +84,7 @@ func RequiredMap[K comparable, T any](_ operation.Context, fldPath *field.Path,
 // type.
 func ForbiddenValue[T comparable](_ operation.Context, fldPath *field.Path, value, _ *T) field.ErrorList {
 	var zero T
-	if value == nil || *value == zero {
+	if *value == zero {
 		return nil
 	}
 	return field.ErrorList{field.Forbidden(fldPath, "")}
diff --git a/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go b/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
index 77f664e5bb2..bfba3269167 100644
--- a/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
+++ b/staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go
@@ -104,15 +104,42 @@ func (forbiddenDeclarativeValidator) ExtractValidations(t *types.Type, comments
 	for t.Kind == types.Alias {
 		t = t.Underlying
 	}
+	// Forbidden is weird.  Each of these emits two checks, which are polar
+	// opposites.  If the field fails the forbidden check, it will
+	// short-circuit and not run the optional check.  If it passes the
+	// forbidden check, it must not be specified, so it will "fail" the
+	// optional check and short-circuit (but without error).  Why?  For
+	// example, this prevents any further validation from trying to run on a
+	// nil pointer.
 	switch t.Kind {
 	case types.Slice:
-		return Validations{Functions: []FunctionGen{Function(forbiddenTagName, ShortCircuit, forbiddenSliceValidator)}}, nil
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenSliceValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalSliceValidator),
+			},
+		}, nil
 	case types.Map:
-		return Validations{Functions: []FunctionGen{Function(forbiddenTagName, ShortCircuit, forbiddenMapValidator)}}, nil
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenMapValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalMapValidator),
+			},
+		}, nil
 	case types.Pointer:
-		return Validations{Functions: []FunctionGen{Function(forbiddenTagName, ShortCircuit, forbiddenPointerValidator)}}, nil
+		return Validations{
+			Functions: []FunctionGen{
+				Function(forbiddenTagName, ShortCircuit, forbiddenPointerValidator),
+				Function(forbiddenTagName, ShortCircuit|NonError, optionalPointerValidator),
+			},
+		}, nil
 	}
-	return Validations{Functions: []FunctionGen{Function(forbiddenTagName, ShortCircuit, forbiddenValueValidator)}}, nil
+	return Validations{
+		Functions: []FunctionGen{
+			Function(forbiddenTagName, ShortCircuit, forbiddenValueValidator),
+			Function(forbiddenTagName, ShortCircuit|NonError, optionalValueValidator),
+		},
+	}, nil
 }
 
 func (forbiddenDeclarativeValidator) Docs() []TagDoc {
```

produces:
```
    // field T1.PT2    
    errs = append(errs,                                                                                                         
        func(obj, oldObj *T2, fldPath *field.Path) (errs field.ErrorList) {    
            if e := validate.ForbiddenPointer(opCtx, fldPath, obj, oldObj); len(e) != 0 {    
                errs = append(errs, e...)    
                return // do not proceed    
            }        
            if e := validate.OptionalPointer(opCtx, fldPath, obj, oldObj); len(e) != 0 {    
                return // do not proceed    
            }    
            errs = append(errs, Validate_T2(opCtx, obj, oldObj, fldPath)...)    
            return    
        }(obj.PT2, safe.Field(oldObj, func(oldObj *T1) *T2 { return oldObj.PT2 }), fldPath.Child("pt2"))...)
```

But now it is weird WRT the struct rule, so maybe we DON'T do the struct rule for forbidden.  The problem is that any struct which holds a slice will fail to compile (not comparable).  So I again offer https://github.com/jpbetz/kubernetes/pull/54#discussion_r1867128162 - don't allow forbidden on non-pointer structs.

Reply:
You can patch that in and try it.

Reply:
Ah.. eventually back to the non-comparable problem.
Ok, I'm leaning to: for non-pointer structs, required/optional/forbidden tags are only for documentation and don't enforce any validation for now. (we defer the validation to their fields)

Reply:
I think it's:

* optional/required: non-pointer structs do not emit any validation
* forbidden: non-pointer structs should produce an error and halt code-gen.  If it ever comes up, we can understand WHY and decide if a limited form of zero-value checking makes sense, or if a pointer type makes sense.


Then I think it is safe to emit both forbidden and optional, with really good comments :)

Reply:
I've updated the code for non-pointer struct:
* optional/required: no validation emit, added comment
* forbidden: halt code-gen, added comment.